File: Erlang/hello_world/src/questions/kex_2/kex_2_battle.erl
-module(kex_2_battle).
-export([game_loop/3]).
-include("../../shared/enemy.hrl").
-include("../../shared/player.hrl").

-spec attack_hit() -> 110.
attack_hit() -> 110.

-spec skill_hit() -> 100.
skill_hit() -> 100.

-spec magic_hit() -> 70.
magic_hit() -> 70.

-spec select_attack(_) -> {'attack', pos_integer(), 110} | {'magic', pos_integer(), 70} | {'skill', pos_integer(), 100}.
select_attack(1) -> {attack, rand:uniform(40) + 60, attack_hit()};
select_attack(2) -> {skill, rand:uniform(100) + 30, skill_hit()};
select_attack(3) -> {magic, rand:uniform(180) + 20, magic_hit()};
select_attack(_) -> select_attack(1).

-spec check_hit_or_miss(_) -> boolean().
check_hit_or_miss(Hit) -> rand:uniform(100) < Hit.

-spec damage_enemy({_, _, _}, #enemy{}) -> #enemy{}.
damage_enemy(AttackMethod, Enemy = #enemy{}) ->
    {_, Dmg, Hit} = AttackMethod,
    HitOrMiss = check_hit_or_miss(Hit),
    if HitOrMiss -> 
        ActualDamage = (if Dmg - Enemy#enemy.defense < 0 -> 0;
                           true -> Dmg - Enemy#enemy.defense
                        end),
        io:format("~Bのダメージ！~n", [ActualDamage]),
        NewEnemyHp = (if Enemy#enemy.hp - ActualDamage < 0 -> 0;
                         true -> Enemy#enemy.hp - ActualDamage
                      end),
        Enemy#enemy{ hp = NewEnemyHp };
       true ->
        io:format("攻撃を外した！~n"),
        Enemy
    end.

-spec damage_player(#enemy{}, #player{}) -> #player{}.
damage_player(Enemy = #enemy{}, Player = #player{}) ->
    HitOrMiss = check_hit_or_miss(Enemy#enemy.hit),
    if HitOrMiss ->
        Injury = (if Enemy#enemy.attack - Player#player.defense < 0 -> 0;
                     true -> Enemy#enemy.attack - Player#player.defense
                  end),
        io:format("~Bのダメージ！~n", [Injury]),
        NewPlayerHp = (if Player#player.hp - Injury < 0 -> 0;
                          true -> Player#player.hp - Injury
                       end),
        Player#player{ hp = NewPlayerHp };
       true ->
        io:format("攻撃を外した！~n"),
        Player
    end.

battle_loop(_ = #enemy{ hp = Hp, name = Name, level = Level }, Player) when Hp =< 0 ->
    io:format("~tsLv.~Bを倒した！~n", [Name, Level]),
    {continue, Player};
battle_loop(Enemy, _ = #player{ hp = Hp }) when Hp =< 0 ->
    io:format("あなたは~tsに負けました！~n", [Enemy#enemy.name]),
    {'end', "ゲームオーバー！"};
battle_loop(Enemy = #enemy{}, Player = #player{}) ->
    io:format("~ts 残りHP：~B~n", [Enemy#enemy.name, Enemy#enemy.hp]),
    {ok, [Choice]} = io:fread("武器を選択してください（１．攻撃　２．特技　３．魔法）＞", "~d"),
    AttackMethod = select_attack(Choice),
    NewEnemy = damage_enemy(AttackMethod, Enemy),
    io:format("~tsの攻撃！~n", [Enemy#enemy.name]),
    NewPlayer = damage_player(Enemy, Player),
    if NewPlayer#player.hp > 0 ->
        io:format("プレイヤー残りHP：~B~n", [NewPlayer#player.hp]),
        battle_loop(NewEnemy, NewPlayer);
       true -> battle_loop(NewEnemy, NewPlayer)
    end.

engage_battle(Enemy = #enemy{}, Player) ->
    io:format("~tsLv.~Bが現れた！~n", [Enemy#enemy.name, Enemy#enemy.level]),
    battle_loop(Enemy, Player).

-spec game_loop(#player{}, integer(), _) -> nonempty_string().
game_loop(Player = #player{}, Kills, Choice) when Choice =:= 0 ->
    "リ〇ミト！\n戦闘回数：" ++ integer_to_list(Kills) ++ "回　残りHP：" ++ integer_to_list(Player#player.hp);
game_loop(Player = #player{ hp = Hp }, Kills, _) when Hp =< 0 -> game_loop(Player, Kills, 0);
game_loop(Player = #player{}, Kills, _) ->
    io:format("~n現HP：~B~n", [Player#player.hp]),
    {ok, [Choice]} = io:fread("奥に進みますか？（１：奥に進む　０．帰る）＞", "~d"),
    case Choice of
        0 -> game_loop(Player, Kills, Choice);
        _ ->
            Ordinal = rand:uniform(3) - 1,
            Enemy = make_enemy(Ordinal),
            Result = engage_battle(Enemy, Player),
            case Result of
                {'end', Message} ->
                    io:format("~ts", [Message]),
                    game_loop(Player#player{ hp = 0 }, Kills, 0);
                {continue, NewPlayer} ->
                    game_loop(NewPlayer, Kills + 1, 1)
            end
    end.

File: Elixir/hello_world_again/lib/kex_2/kex_2_battle.ex
defmodule Kex2_Battle do
  @spec select_attack(any) ::
          {:attack, number, 110} | {:magic, number, 70} | {:skill, number, 100}
  def select_attack(1), do: {:attack, Enum.random(0..40) + 60, Constants.attack_hit}
  def select_attack(2), do: {:skill, Enum.random(0..100) + 30, Constants.skill_hit}
  def select_attack(3), do: {:magic, Enum.random(0..180) + 20, Constants.magic_hit}
  def select_attack(_), do: select_attack(1)

  @spec check_hit_or_miss(any) :: boolean
  def check_hit_or_miss(hit), do: Enum.random(0..100) < hit

  def damage_enemy(attack_method, enemy) do
    {_, dmg, hit} = attack_method
    hit_or_miss = check_hit_or_miss(hit)
    if hit_or_miss do
      actual_damage = if dmg - enemy.defense < 0 do
        0
      else
        dmg - enemy.defense
      end
      IO.puts("#{actual_damage}のダメージ！")
      new_enemy_hp = if enemy.hp - actual_damage < 0 do
        0
      else
        enemy.hp - actual_damage
      end
      Enemy.new(enemy.level, enemy.name, new_enemy_hp, enemy.defense, enemy.attack, enemy.flee, enemy.hit)
    else
      IO.puts("攻撃を外した！")
      enemy
    end
  end

  def damage_player(enemy, player) do
    hit_or_miss = check_hit_or_miss(enemy.hit)
    if hit_or_miss do
      injury = if enemy.attack - player.defense < 0 do
        0
      else
        enemy.attack - player.defense
      end
      IO.puts("#{injury}のダメージ！")
      new_player_hp = if player.hp - injury < 0 do
        0
      else
        player.hp - injury
      end
      Player.new(new_player_hp, player.defense)
    else
      IO.puts("攻撃を外した！")
      player
    end
  end

  def battle_loop(enemy, player) when enemy.hp <= 0 do
    IO.puts("#{enemy.name}Lv.#{enemy.level}を倒した！")
    {:continue, player}
  end

  def battle_loop(enemy, player) when player.hp <= 0 do
    IO.puts("あなたは#{enemy.name}に負けました！")
    {:end, "ゲームオーバー！"}
  end

  def battle_loop(enemy, player) do
    IO.puts("#{enemy.name} 残りHP：#{enemy.hp}")
    choice = IO.gets("武器を選択してください（１．攻撃　２．特技　３．魔法）＞")
    |> String.trim()
    |> String.to_integer()
    attack_method = select_attack(choice)
    new_enemy = damage_enemy(attack_method, enemy)
    IO.puts("#{enemy.name}の攻撃！")
    new_player = damage_player(enemy, player)
    if new_player.hp > 0 do
      IO.puts("プレイヤー残りHP：#{new_player.hp}")
      battle_loop(new_enemy, new_player)
    else
      battle_loop(new_enemy, new_player)
    end
  end

  def engage_battle(enemy, player) do
    IO.puts("#{enemy.name}Lv.#{enemy.level}が現れた！")
    battle_loop(enemy, player)
  end

  def game_loop(player, kills, choice) when choice == 0 or player.hp <= 0 do
    "リ〇ミト！\n戦闘回数：#{kills}回　残りHP：#{player.hp}"
  end
  def game_loop(player, kills, _) do
    IO.puts("\n現HP：#{player.hp}")
    choice = IO.gets("奥に進みますか？（１：奥に進む　０．帰る）＞")
    |> String.trim()
    |> String.to_integer()
    case choice do
      0 ->
        game_loop(player, kills, choice)
      _ ->
        ordinal = Enum.random(0..2)
        {_, enemy} = Enemy.make_enemy(ordinal)
        case engage_battle(enemy, player) do
          {:end, s} ->
            IO.puts(s)
            game_loop(Player.new(0, player.defense), kills, 0)
          {:continue, p} ->
            game_loop(p, kills + 1, 1)
        end
    end
  end
end
